site/book/04-using-functions/01-declarative-function-execution.md

```diff
@@ -0,0 +1,179 @@
+
+In kpt, this is achieved by declaring a pipleline of functions in the `Kptfile` and executing
+all the pipelines in the package hierarchy in a depth-first order using the `fn render` command.
+
+In our wordpress example, the top-level `wordpress` package declares this pipeline:
+
+```yaml
+# wordpress/Kptfile (Excerpt)
+apiVersion: kpt.dev/v1alpha2
+kind: Kptfile
+metadata:
+  name: wordpress
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configMap:
+        wp-image: wordpress
+        wp-tag: 4.8-apache
+  validators:
+    - image: gcr.io/kpt-fn/kubeval:v0.1
+```
+
+This declares two functions:
+
+- `apply-setters` is an example mutator function which provides a form of parameterization. You
+  don't need to understand the nitty-gritty of how this function works except that it mutates
+  resources that have been parameterized with a given key (e.g. `wp-tag`) by applying the
+  corresponding value (e.g. `4.8-apache`).
+- `kubeval` is an example validator function which you've seen before. It validates the resources
+  against their OpenAPI schema.
+
+> Refer to the [Functions Catalog][func-cat] for details on how to use a particular function.
+
+There are two differences between mutators and validators:
+
+1. Validators are not allowed to modify resources.
+2. Validators are always executed after mutators.
+
+The `mysql` subpackage declares only a mutator function:
+
+```yaml
+# wordpress/mysql/Kptfile
+apiVersion: kpt.dev/v1alpha2
+kind: Kptfile
+metadata:
+  name: mysql
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configMap:
+        ms-image: mysql
+        ms-tag: 5.6
+```
+
+Now, let's render the package hierarchy:
+
+```shell
+$ kpt fn render wordpress
+Package "wordpress/mysql":
+
+[PASS] "gcr.io/kpt-fn/apply-setters:v0.1"
+
+Package "wordpress":
+
+[PASS] "gcr.io/kpt-fn/apply-setters:v0.1"
+[PASS] "gcr.io/kpt-fn/kubeval:v0.1"
+
+Successfully executed 3 function(s) in 2 package(s).
+```
+
+When you invoke the `render` command, kpt performs the following steps:
+
+1. Sequentially executes the list of mutators declared in the `mysql` package. The input to the
+   first function is the set of resources read from the configuration files in the `mysql` package.
+   The output of the first function is the input of the second function and so on.
+2. Similarly, executes all the validators declared in the `mysql` package. The input to the first
+   validator is the output of the last mutator. The output of the last validator is the
+   output of the pipeline in the `mysql` package.
+3. Sequentially executes the list of mutators declared in the `wordpress` package. The input to the
+   first function is the union of:
+
+   - Resources read from configuration files in the `wordpress` package AND
+   - Output of the pipeline from the `mysql` package (Step 2).
+
+     Note that it is possible for a mutator in `wordpress` to modify resources in the `mysql` subpackage.
```

Comment:
NB: possible to outline here that the parent/child relationship is the reason.  It's implied but might be worth reiterating.